x/bank/keeper/keeper.go

```diff
@@ -104,12 +108,31 @@ func NewBaseKeeper(
 	}
 
 	return BaseKeeper{
-		BaseSendKeeper: NewBaseSendKeeper(cdc, storeKey, ak, paramSpace, blockedAddrs),
-		ak:             ak,
-		cdc:            cdc,
-		storeKey:       storeKey,
-		paramSpace:     paramSpace,
+		BaseSendKeeper:         NewBaseSendKeeper(cdc, storeKey, ak, paramSpace, blockedAddrs),
+		ak:                     ak,
+		cdc:                    cdc,
+		storeKey:               storeKey,
+		paramSpace:             paramSpace,
+		mintCoinsRestrictionFn: func(ctx sdk.Context, coins sdk.Coins) error { return nil },
+	}
+}
+
+// WithMintCoinsRestriction restricts the bank Keeper used within a specific module to
+// have restricted permissions on minting speicified denoms.
+func (k BaseKeeper) WithMintCoinsRestriction(NewRestrictionFn BankMintingRestrictionFn) BaseKeeper {
+	oldRestrictionFn := k.mintCoinsRestrictionFn
+	k.mintCoinsRestrictionFn = func(ctx sdk.Context, coins sdk.Coins) error {
+		err := NewRestrictionFn(ctx, coins)
```

Comment:
Wait, why are we calling both old and new restriction functions?

Reply:
We want it to nest. So I can do `bankKeeper.WithMintCoinsRestriction(restriction1).WithMintCoinsRestriction(restriction2)`

Reply:
Ok the godoc should clearly reflect this, because looking at the method, that's not clear at all.

Reply:
And I'm not even sure this is the best approach. With this approach, you don't have a clear way of reseting it. Also, you could just instead create a single restriction function that nests all of the "inner" ones.

Reply:
Why do you need to reset it though? Idea is to give a separate keeper instance to each module in its init. E.g. in Osmosis for the Gamm keepers initialization: https://github.com/osmosis-labs/osmosis/blob/main/app/keepers.go#L205-L208

I want to change that to:
```
	gammKeeper := gammkeeper.NewKeeper(
		appCodec, keys[gammtypes.StoreKey],
		app.GetSubspace(gammtypes.ModuleName),
		app.AccountKeeper, 
		app.BankKeeper.WithMintCoinsRestriction(NewPrefixMintCoinsRestrictionFn("gamm/pool/")), 
		app.DistrKeeper)
		```

Reply:
(I'm also down for a reset method if thats helpful though!)

Reply:
In that case you're just providing one function. Where do you nest them? I just don't think this approach is clean. But I'm not suggesting we block on it.

Reply:
Yeah so in my alternative proposal (https://github.com/cosmos/cosmos-sdk/pull/10771#discussion_r779142999), you could easily create a function that calls _both_ `NewBlockPrefixMintCoinsRestrictionFn` and `NewBlockPrefixMintCoinsRestrictionFn`.

```go
func Foo() {
 if err := NewBlockPrefixMintCoinsRestrictionFn(); err !=nil {
  return err
 }

// ...
}
```

Reply:
@ValarDragon proposes a middleware approach, so if we have a bank keeper with default setting we make sure the restriction fn will always be called. 
With @alexanderbez  approach we need to remember to put include the default check function in all custom functions.